segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * An in-memory implementation of cache for read index that maps chunk start offset to chunk name for recently used segments.
+ * Eviction is performed only when number of segments or chunks exceeds certain given limits.
+ * Each time items are evicted the generation is incremented.
+ * For each entry in cache, we keep track of the generation in which it was last accessed.
+ * During eviction items with oldest generation are evicted first until enough objects are evicted.
+ * The least accessed segments are removed entirely before removing chunks from more recently used segments.
+ * This calculation is "best effort" and need not be accurate.
+ */
+class ReadIndexCache {
+
+    /**
+     * Max number of indexed segments to keep in cache.
+     */
+    private final int maxIndexedSegments;
+
+    /**
+     * Max number of indexed chunks to keep in cache.
+     */
+    private final int maxIndexedChunks;
+
+    /**
+     * Max number of indexed chunks to keep per segment in cache.
+     */
+    private final int maxIndexedChunksPerSegment;
+
+    /**
+     * Current generation of cache entries.
+     */
+    private final AtomicLong currentGeneration = new AtomicLong();
+
+    /**
+     * Lowest generation of cache entries.
+     */
+    private final AtomicLong oldestGeneration = new AtomicLong();
+
+    /**
+     * Total number of chunks in the cache.
+     */
+    private final AtomicInteger totalChunkCount = new AtomicInteger();
+
+    /**
+     * Index of chunks for a segment by their start offsets.
+     */
+    private final ConcurrentHashMap<String, SegmentReadIndex> segmentsToReadIndexMap = new ConcurrentHashMap<String, SegmentReadIndex>();
```

Comment:
Can you justify the use of `ConcurrentHashMap` vs. other data structure implementations, e.g., `LinkedHashMap`? I'm thinking that it would avoid the whole management of generations and a full iteration over the map upon eviction.

Reply:
Conceptually - as entries are found and used, they need to be moved to the end of list to keep LRU order.
`LinkedHashMap` keeps order by of insertion order, which is different than what we want.

Reply:
Resolving as per discussion offline. Tracking this here https://github.com/pravega/pravega/issues/4902